### ISSUE TITLE
Suspicious unwrap in Account's deserialize

### ISSUE THREAD

**Issue by matklad** [near/nearcore#6186]
The following code looks somewhat suspicious to me: 

https://github.com/nearprotocol/nearcore/blob/47620d3ec7cdd0d9c30d981d537234ffeb8c0b47/core/primitives-core/src/account.rs#L114-L130

`Borsh` generally doesn't have a contract that the input is valid, and I think we should return a borsh error in this case. 

That code is reachable from 

https://github.com/nearprotocol/nearcore/blob/47620d3ec7cdd0d9c30d981d537234ffeb8c0b47/core/store/src/lib.rs#L334-L340

which converts borsh serialization errors to inconsistent state errors, which sounds like the right logic here.

**Comment by matklad** [comment]
@EgorKulikov could you take a look when you have time?

**Comment by EgorKulikov** [comment]
Yep, I agree

**Comment by Ekleog** [comment]
Overall one thing that bothers me with this assertion is that borsh appears to have as an expectation that BorshDeserialize will advance the buffer by the number of bytes consumed, which sounds to me like it could always pass in a slice bigger than the actual data to deserialize. So I'm not sure even returning an error would actually be implementing the BorshDeserialize as specified.

But then I'm having doubts, because in practice it seems like the code we currently have, that panics on error, does work, so… maybe I'm misunderstanding the BorshDeserialize contract?

**Comment by frol** [comment]
Should this issue be closed as #6388 was merged?

**Comment by matklad** [comment]
No, the code still violates the contract of BorshDeserialize trait, by panicing on slices of invalid lenght

**Comment by frol** [comment]
I would just turn the panic into a warning log. WDYT?

**Comment by Ekleog** [comment]
I'm not sure it'd be much better than the statu quo. The reason for removing the panic is (IMO) mostly to simplify code, as if it were actually reachable there's a good chance that it'd already have triggered. Plus primitives-core doesn't depend on any logging crate yet, so I'm having doubts warn-logging would be of much help here.

That said, it'd be different if we actually thought there's a chance that the panic actually triggers, in which case we would have to think about whether this is:
- expected behavior (which I think, as it's in the documented BorshDeserialize API), in which case warn-logging would be surprising for the end-user as there's actually nothing wrong to warn about
- unexpected behavior, in which case we may have just deserialized an invalid account and crashing might be a better choice than continuing with broken data

To sum up, the decision tree as I see it:
- Do we expect it to ever trigger?
  - No -> can either keep the panic, remove it or warn-log, it doesn't matter. Code is simpler with removing everything, and warn-logging would add a dep and so add a bit of complexity
  - Yes -> Would that be expected behavior within the bounds of API usage?
    - Yes -> We should remove the panic and not have warn-logging as it's normal operation
    - No -> We should keep the panic so as not to risk continuing with corrupted data

WDYT?

**Comment by frol** [comment]
> Do we expect it to ever trigger?

No, it is definitely not expected to ever get triggered. Given that indeed this code was "tested in production" we can probably just drop the condition and the body.

**Comment by Ekleog** [comment]
Got it! The PR for doing so is at #7013 then :)
